### PR TITLE
fix(security): strip authToken cookie before forwarding to project code

### PR DESCRIPTION
**Pull request by kojiwakayama**
## Summary

The hosted proxy consumes the `authToken` cookie (via `extractUserToken` in `src/proxy/proxy-token-resolution.ts`) to resolve the caller's Veryfront identity, but `createProxyContextHeaders` then forwarded the Cookie header verbatim to the deployed project — it only removed internal `x-*` proxy headers, `host`, and hop-by-hop fields. Any tenant-controlled page, API route, or middleware in a protected environment could therefore read the raw Veryfront credential of anyone who visited it. This includes the deployer's account session JWT: the CLI's post-deploy readiness probe (`waitForEnvironmentReady` in `cli/shared/deployment/deploy-project.ts`) sends `Cookie: authToken=<token>` to a normal app route, and for `veryfront login` session credentials no environment-scoped token exchange applies (only opaque API keys were exchanged, per #3959/#3966). A malicious project collaborator or compromised dependency could exfiltrate the deployer's live full-account credential during every deploy — a real privilege escalation across the platform/tenant boundary.

## Finding

- Codex finding id: `67637779a99c8191877f33f62588ec2b` — severity: high
- Introduced in cdccb25a76aa86f86fb1eaf3a3d67414af10aa5e

## Fix

`createProxyContextHeaders` (`src/proxy/handler.ts`) now strips every `authToken` pair from the Cookie header before building the downstream request, dropping the header entirely when no other cookies remain. A new `stripUserTokenCookie` helper lives next to `extractUserToken` so consumption and redaction stay in one place. The proxy already forwards the resolved identity via `x-token`, so nothing downstream loses information, and application cookies pass through unchanged. All downstream forwarding paths (HTTP forward, split forward, WebSocket bridge) share `createProxyContextHeaders`, so each is covered. The only other `authToken` cookie consumer (`src/agent/service/auth.ts`) runs as a standalone service on its own port and also accepts `Authorization: Bearer`, so it is unaffected.

Defense-in-depth follow-up (not in this PR): extend the CLI's `resolveEnvironmentAccess` exchange to also cover session JWTs so the readiness probe only ever presents a short-lived environment-bound token; requires control-plane support for exchanging session credentials.

## Test evidence

- New unit tests: `stripUserTokenCookie` edge cases (multiple pairs, whitespace, bare `authToken`, name lookalikes preserved) in `src/proxy/proxy-token-resolution.test.ts`; `injectContextHeaders` strips the `authToken` cookie while preserving app cookies and drops the header when it was the only cookie, in `src/proxy/handler.test.ts`.
- `deno task test:file` green on: `src/proxy/proxy-token-resolution.test.ts` (7 steps), `src/proxy/handler.test.ts` (75 steps), `src/proxy/split-forward-request.test.ts`, `src/proxy/proxy-access-control.test.ts`, `src/proxy/websocket-proxy.test.ts`, `src/proxy/mode-parity.test.ts`, `src/proxy/websocket-bridge-identity.test.ts`, `src/proxy/hop-by-hop-headers.test.ts` — 0 failures.
- `deno fmt --check`, `deno lint`, and `deno check` clean on all touched files.

https://claude.ai/code/session_01QfWNMiUhvWMKWi6BGfVdY3

## Summary by CodeRabbit

- **Bug Fixes**
  - Hosted project requests no longer forward the platform authentication cookie.
  - Other cookies remain intact, while the cookie header is removed when no cookies remain.
  - Local project requests continue preserving application authentication cookies.

- **Tests**
  - Added coverage for mixed, token-only, whitespace, valueless, and similarly named cookies.

### CODE REVIEW

**Comment by kwakayama** [comment]
## Code Review: 92/100 — Excellent

Well-scoped fix for a real high-severity credential-leak, applied at the correct choke point and backed by solid tests.

**Strengths**
- Correctly closes a genuine vulnerability: the raw `authToken` cookie (including the deployer's full-account session JWT sent by the CLI's readiness probe) was forwarded verbatim to tenant-controlled project code. The fix lives in `createProxyContextHeaders`, the single function shared by all three forwarding paths (HTTP forward, split forward, WebSocket bridge — confirmed via `injectContextHeaders` and direct callers in `split-forward-request.ts` / `websocket-bridge.ts`), so no path is left exposed.
- `stripUserTokenCookie` strips *every* `authToken` pair, not just the first — good defense-in-depth given `extractUserToken` bails out entirely on duplicate `authToken` pairs (ambiguous case). Non-matching cookies, including name-lookalikes (`authTokenX`, `XauthToken`), are correctly preserved verbatim so application cookies still work.
- Test coverage is targeted and covers the edges that matter: unit tests for the helper (multiple pairs, whitespace, bare `authToken`, empty header, lookalike names) plus integration-level tests confirming `injectContextHeaders` both preserves app cookies and fully drops the `Cookie` header when nothing remains.
- Comments explain *why* (identity now flows via `x-token`) rather than just what, which will help the next person who touches this function.

**Minor notes (non-blocking)**
- `stripUserTokenCookie` re-implements the same cookie-pair split/trim logic already in `extractUserToken`. Not a defect given both are small and independently well-tested, but a shared low-level splitter would remove the duplication.
- The PR body mentions a defense-in-depth follow-up (extending `resolveEnvironmentAccess` to cover session JWTs) as explicitly out of scope — worth making sure that actually gets tracked as a follow-up issue rather than just living in this PR's description.

No functional or security concerns with the change as scoped. Nice, surgical fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/proxy/handler.ts [hunk 1]
```diff
@@ -1367,6 +1371,18 @@ export function createProxyContextHeaders(
...
   for (const header of INTERNAL_PROXY_HEADERS) headers.delete(header);
 
+  // The `authToken` cookie carries the caller's Veryfront credential (a user
+  // session JWT, or an exchanged environment access token). The proxy has
+  // already consumed it (resolveProxyRequestToken) and forwards the resolved
+  // identity via `x-token`; the raw credential must never reach
+  // tenant-controlled project code. Application cookies pass through intact.
+  const cookieHeader = headers.get("cookie");
+  if (cookieHeader !== null) {
+    const remainingCookies = stripUserTokenCookie(cookieHeader);
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/proxy/handler.ts]: Fixed in 20cb704793116c8bbe6fa07b478d30cc5a8a480f. Local projects now bypass authToken redaction because processRequest already marks them isLocalProject and skips resolveProxyRequestToken; hosted/non-local requests retain the existing platform-cookie stripping. Added a regression that exercises localProjects with proxy client credentials, confirms no proxy token is consumed, and verifies the application authToken cookie is forwarded intact. Verification (Deno 2.7.7): deno task test:file src/proxy/token-priority.test.ts src/proxy/handler.test.ts (2 passed, 91 steps); deno fmt --check and deno lint on both changed files; deno check src/proxy/handler.ts. The full pre-push gate also cleared formatting, lint, typecheck, and the parallel unit suite before an unrelated serial esbuild test harness ended with a pending-promise runner error.

**Review by kwakayama — COMMENTED** [review]
### Findings

- [P2] `src/proxy/handler.ts:1379-1385` strips every `authToken` cookie for every non-local project, including custom domains, while `src/proxy/proxy-token-resolution.ts:96-107` has no way to distinguish a platform credential from an application’s same-named session cookie. Any hosted application using `authToken` for its own authentication will silently lose that cookie. The new local-only exemption (`src/proxy/token-priority.test.ts:381-410`) demonstrates this collision is real, but hosted deployments receive no equivalent compatibility path, migration guidance, or reserved-cookie documentation. Use a distinguishable platform credential namespace or document and deliberately manage the breaking reservation.

- [P3] The WebSocket coverage is now misleading and does not prove the security boundary. `src/proxy/websocket-client.test.ts:175-194` sends `authToken=browser-session` and says cookies survive the bridge, but the fixture records no cookie (`src/proxy/websocket-client.test.ts:35-40`) and assertions at `:219-221` never inspect it. Although the bridge uses the shared helper (`src/proxy/websocket-bridge.ts:69`), add an upstream assertion that `authToken` is absent while an unrelated application cookie remains present.

| Area | Score |
| --- | ---: |
| Correctness | 34/40 |
| Tests | 13/20 |
| Reliability/security | 15/15 |
| Maintainability | 14/15 |
| Scope/docs | 5/10 |
| **Total** | **81/100** |

Review-Gate:
Reviewer: Codex
Reviewed-SHA: 20cb704793116c8bbe6fa07b478d30cc5a8a480f
Score: 81/100
Actionable-Findings: 2
Verdict: REQUEST_CHANGES

**Comment by kojiwakayama** [comment]
@codex review
